sanitycheck: add west-runner parameter and fix-up west-flash handling

mike-scott opened the pull request:
This patch series does the following:
- Several boards have multiple runners.  We need a way to specify which runner to use with sanitycheck.  Introduce --west-runner option for that purpose.
- Generate an error when --west-runner is used w/o the --west-flash option.
- Cleans up the help text of --west-flash.
- Generate an error when --west-flash is used w/o the --device-testing option.

mike-scott committed:
sanitycheck: add west-runner parameter

Several boards have multiple runners setup.  We need a way to specify
which runner to use with sanitycheck.  Introduce --west-runner option.

Signed-off-by: Michael Scott <mike@foundries.io>

nashif commented on scripts/sanitycheck:
```diff
@@ -676,6 +676,9 @@ class DeviceHandler(Handler):
         if options.west_flash is not None:
             command = ["west", "flash", "--skip-rebuild", "-d", self.outdir]
+            if options.west_runner is not None:
+                command.append("-r")
```
prefer "--runner" instead of "-r", self documenting.

mike-scott replied:
good point.  fixing.

marc-hb commented on scripts/sanitycheck:
```diff
@@ -3021,6 +3024,17 @@ Artificially long but functional example:
+                     --west-runner="pyocd"
+        will translate to
+         west flash -r pyocd
+        """
```
Based on the code change above, `--west-runner` seems silently ignored when `--west-flash` is missing. However this example in the help doesn't use `--west-flash`, only `--west-runner`?

If `--west-flash` is really required to use `--west-runner` then instead of dropping it on the floor add a warning or maybe even an error if someone or some buggy script makes the mistake of using the latter without the former.

mike-scott replied:
west-flash is indeed required for west-runner.  added a check to ensure this was true and generate an error if not.

marc-hb replied:
Late suggestion sorry: another, possibly more user-friendly way would be to automatically set `argparse.west_flash=[]` when `--west-runner` is present and `--west-flash` is missing (and of course say so in the --help). It would save the user some typing. Would it make sense from a `west` point of view? I mean:
- Does `--west-runner` make sense with a default `--west-flash` without any explicit value?
- is there really no way some uneducated user could imagine (and be wrong) that `--west-runner` makes sense without any `--west-flash` at all?

mike-scott replied:
As this is a flashing command, I would rather set any defaults if the user sets `--west-runner` without enabling `--west-flash`.  I'm open to suggestions here tho.  That's just my opinion.

marc-hb reviewed (changes requested):
As these options seem at least related to each other or maybe even one dependent on the other, add a new argument group, it's very simple:  https://pymotw.com/2/argparse/#argument-groups

mike-scott committed:
sanitycheck: handle west-flash dependency for device-testing

sanitycheck --help mentions that west-flash requires device-testing
to be enabled (and indeed it does because DeviceHandler will never
be called where west_flash option is used.  Let's generate an error
if west-flash is used w/o specifying device-testing.

Also cleanup help text which looks odd in both sanitycheck --help
and in the file itself.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@marc-hb please re-review.  I've addressed your issues, with the exception of splitting out west-flash and west-runner into an argument group.  I didn't like how it was rendering them as a separate group compared to all of the other options.  Looked very odd.  If this is needed, we should establish a new PR for that specific purpose: to break out *all* of the options into their respective groups.  There will probably be discussion on group naming, etc.

mike-scott commented on scripts/sanitycheck:
```diff
@@ -3021,6 +3024,17 @@ Artificially long but functional example:
         """
     )
+    parser.add_argument(
+        "--west-runner", nargs='?', const=[],
```
it cannot be used bare.  removed erroneously copied nargs/const params.

marc-hb reviewed (approved):
Approving from an `argparse` perspective however I don't know `west` enough west to comment about it.

> I didn't like how it was rendering them as a separate group compared to all of the other options. Looked very odd. If this is needed, we should establish a new PR for that specific purpose: to break out all of the options into their respective groups. There will probably be discussion on group naming, etc.

There's at least one group already. It didn't require any long naming discussion or anything. Funny how you didn't notice it despite how odd it looks ;-)

mike-scott commented:
> There's at least one group already. It didn't require any long naming discussion or anything. Funny how you didn't notice it despite how odd it looks ;-)

Sarcasm noted.

It's really a personal preference but here's the output of sanitycheck for this PR:
```
$ sanitycheck --help
usage: sanitycheck [-h] [--force-toolchain] [-p PLATFORM] [-a ARCH] [-t TAG]
                   [-e EXCLUDE_TAG] [-f] [-c CONFIG]
                   [-s TEST | --sub-test SUB_TEST] [-l] [-o TESTCASE_REPORT]
                   [-d DISCARD_REPORT] [--compare-report COMPARE_REPORT]
                   [-B SUBSET] [-N] [-y] [--list-tags] [--report-excluded]
                   [--list-tests] [--export-tests FILENAME]
                   [--detailed-report FILENAME] [--timestamps] [-r] [-w] [-v]
                   [-i] [--log-file FILENAME] [-m] [-u] [-F FILENAME]
                   [-E FILENAME] [-b | --test-only] [--cmake-only] [-j JOBS]
                   [--device-testing] [-X FIXTURE]
                   [--device-serial DEVICE_SERIAL] [--show-footprint]
                   [-H FOOTPRINT_THRESHOLD] [-D] [-O OUTDIR] [-n]
                   [-T TESTCASE_ROOT] [-A BOARD_ROOT] [-z SIZE] [-S]
                   [--disable-unrecognized-section-test] [-R]
                   [--disable-asserts] [-Q] [--disable-size-report]
                   [-x EXTRA_ARGS] [--west-flash [WEST_FLASH]]
                   [--west-runner WEST_RUNNER] [--gcov-tool GCOV_TOOL]
                   [--enable-coverage] [-C]
                   [--coverage-platform COVERAGE_PLATFORM]

Zephyr Sanity Tests

Also check the "User and Developer Guides" at https://docs.zephyrproject.org/

This script scans for the set of unit test applications in the git
repository and attempts to execute them. By default, it tries to
build each test case on one platform per architecture, using a precedence
list defined in an architecture configuration file, and if possible
run the tests in the QEMU emulator.

Test cases are detected by the presence of a 'testcase.yaml' or a sample.yaml
files in the application's project directory. This file may contain one or more
blocks, each identifying a test scenario. The title of the block is a name for
the test case, which only needs to be unique for the test cases specified in
that testcase meta-data. The full canonical name for each test case is <path to
test case>/<block>.

Each test block in the testcase meta data can define the following key/value
pairs:

  tags: <list of tags> (required)
    A set of string tags for the testcase. Usually pertains to
    functional domains but can be anything. Command line invocations
    of this script can filter the set of tests to run based on tag.

  skip: <True|False> (default False)
    skip testcase unconditionally. This can be used for broken tests.

  slow: <True|False> (default False)
    Don't build or run this test case unless --enable-slow was passed
    in on the command line. Intended for time-consuming test cases
    that are only run under certain circumstances, like daily
    builds.

  extra_args: <list of extra arguments>
    Extra cache entries to pass to CMake when building or running the
    test case.

  extra_configs: <list of extra configurations>
    Extra configuration options to be merged with a master prj.conf
    when building or running the test case.

  build_only: <True|False> (default False)
    If true, don't try to run the test even if the selected platform
    supports it.

  build_on_all: <True|False> (default False)
    If true, attempt to build test on all available platforms.

  depends_on: <list of features>
    A board or platform can announce what features it supports, this option
    will enable the test only those platforms that provide this feature.

  min_ram: <integer>
    minimum amount of RAM needed for this test to build and run. This is
    compared with information provided by the board metadata.

  min_flash: <integer>
    minimum amount of ROM needed for this test to build and run. This is
    compared with information provided by the board metadata.

  timeout: <number of seconds>
    Length of time to run test in QEMU before automatically killing it.
    Default to 60 seconds.

  arch_whitelist: <list of arches, such as x86, arm, arc>
    Set of architectures that this test case should only be run for.

  arch_exclude: <list of arches, such as x86, arm, arc>
    Set of architectures that this test case should not run on.

  platform_whitelist: <list of platforms>
    Set of platforms that this test case should only be run for.

  platform_exclude: <list of platforms>
    Set of platforms that this test case should not run on.

  extra_sections: <list of extra binary sections>
    When computing sizes, sanitycheck will report errors if it finds
    extra, unexpected sections in the Zephyr binary unless they are named
    here. They will not be included in the size calculation.

  filter: <expression>
    Filter whether the testcase should be run by evaluating an expression
    against an environment containing the following values:

    { ARCH : <architecture>,
      PLATFORM : <platform>,
      <all CONFIG_* key/value pairs in the test's generated defconfig>,
      <all DT_* key/value pairs in the test's generated device tree file>,
      <all CMake key/value pairs in the test's generated CMakeCache.txt file>,
      *<env>: any environment variable available
    }

    The grammar for the expression language is as follows:

    expression ::= expression "and" expression
                 | expression "or" expression
                 | "not" expression
                 | "(" expression ")"
                 | symbol "==" constant
                 | symbol "!=" constant
                 | symbol "<" number
                 | symbol ">" number
                 | symbol ">=" number
                 | symbol "<=" number
                 | symbol "in" list
                 | symbol ":" string
                 | symbol

    list ::= "[" list_contents "]"

    list_contents ::= constant
                    | list_contents "," constant

    constant ::= number
               | string

    For the case where expression ::= symbol, it evaluates to true
    if the symbol is defined to a non-empty string.

    Operator precedence, starting from lowest to highest:

        or (left associative)
        and (left associative)
        not (right associative)
        all comparison operators (non-associative)

    arch_whitelist, arch_exclude, platform_whitelist, platform_exclude
    are all syntactic sugar for these expressions. For instance

        arch_exclude = x86 arc

    Is the same as:

        filter = not ARCH in ["x86", "arc"]

    The ':' operator compiles the string argument as a regular expression,
    and then returns a true value only if the symbol's value in the environment
    matches. For example, if CONFIG_SOC="quark_se" then

        filter = CONFIG_SOC : "quark.*"

    Would match it.

The set of test cases that actually run depends on directives in the testcase
filed and options passed in on the command line. If there is any confusion,
running with -v or --discard-report can help show why particular test cases
were skipped.

Metrics (such as pass/fail state and binary size) for the last code
release are stored in scripts/sanity_chk/sanity_last_release.csv.
To update this, pass the --all --release options.

To load arguments from a file, write '+' before the file name, e.g.,
+file_name. File content must be one or more valid arguments separated by
line break instead of white spaces.

Most everyday users will run with no arguments.

optional arguments:
  -h, --help            show this help message and exit
  --force-toolchain     Do not filter based on toolchain, use the set
                        toolchain unconditionally
  -p PLATFORM, --platform PLATFORM
                        Platform filter for testing. This option may be used
                        multiple times. Testcases will only be built/run on
                        the platforms specified. If this option is not used,
                        then platforms marked as default in the platform
                        metadata file will be chosen to build and test.
  -a ARCH, --arch ARCH  Arch filter for testing. Takes precedence over
                        --platform. If unspecified, test all arches. Multiple
                        invocations are treated as a logical 'or' relationship
  -t TAG, --tag TAG     Specify tags to restrict which tests to run by tag
                        value. Default is to not do any tag filtering.
                        Multiple invocations are treated as a logical 'or'
                        relationship
  -e EXCLUDE_TAG, --exclude-tag EXCLUDE_TAG
                        Specify tags of tests that should not run. Default is
                        to run all tests with all tags.
  -c CONFIG, --config CONFIG
                        Specify platform configuration values filtering. This
                        can be specified two ways: <config>=<value> or just
                        <config>. The defconfig for all platforms will be
                        checked. For the <config>=<value> case, only match
                        defconfig that have that value defined. For the
                        <config> case, match defconfig that have that value
                        assigned to any value. Prepend a '!' to invert the
                        match.
  -l, --all             Build/test on all platforms. Any --platform arguments
                        ignored.
  -o TESTCASE_REPORT, --testcase-report TESTCASE_REPORT
                        Output a CSV spreadsheet containing results of the
                        test run. The handler_time column is left blank for
                        tests that were only compiled and not run.
  -d DISCARD_REPORT, --discard-report DISCARD_REPORT
                        Output a CSV spreadsheet showing tests that were
                        skipped and why
  --compare-report COMPARE_REPORT
                        Use this report file for size comparison
  -B SUBSET, --subset SUBSET
                        Only run a subset of the tests, 1/4 for running the
                        first 25%, 3/5 means run the 3rd fifth of the total.
                        This option is useful when running a large number of
                        tests on different hosts to speed up execution time.
  -N, --ninja           Use the Ninja generator with CMake
  -y, --dry-run         Create the filtered list of test cases, but don't
                        actually run them. Useful if you're just interested in
                        --discard-report
  --list-tags           list all tags in selected tests
  --report-excluded     List all tests that are never run based on current
                        scope and coverage. If you are looking for accurate
                        results, run this with --all, but this will take a
                        while...
  --export-tests FILENAME
                        Export tests case meta-data to a file in CSV format.
  --detailed-report FILENAME
                        Generate a junit report with detailed testcase
                        results. Unlike the CSV file produced by --testcase-
                        report, this XML report includes only tests which have
                        run and none which were merely built. If an image with
                        multiple tests crashes early then later tests are not
                        accounted for either.
  --timestamps          Print all messages with time stamps
  -r, --release         Update the benchmark database with the results of this
                        test run. Intended to be run by CI when tagging an
                        official release. This database is used as a basis for
                        comparison when looking for deltas in metrics such as
                        footprint
  -w, --warnings-as-errors
                        Treat warning conditions as errors
  -v, --verbose         Emit debugging information, call multiple times to
                        increase verbosity
  -i, --inline-logs     Upon test failure, print relevant log data to stdout
                        instead of just a path to it
  --log-file FILENAME   log also to file
  -m, --last-metrics    Instead of comparing metrics from the last --release,
                        compare with the results of the previous sanity check
                        invocation
  -u, --no-update       do not update the results of the last run of the
                        sanity checks
  -b, --build-only      Only build the code, do not execute any of it in QEMU
  --test-only           Only run device tests with current artifacts, do not
                        build the code
  --cmake-only          Test on device directly. Specify the serial device to
                        use with the --device-serial option.
  -j JOBS, --jobs JOBS  Number of jobs for building, defaults to number of CPU
                        threads, overcommited by factor 2 when --build-only
  --device-testing      Test on device directly. Specify the serial device to
                        use with the --device-serial option.
  -X FIXTURE, --fixture FIXTURE
                        Specify a fixture that a board might support
  --device-serial DEVICE_SERIAL
                        Serial device for accessing the board (e.g.,
                        /dev/ttyACM0)
  --show-footprint      Show footprint statistics and deltas since last
                        release.
  -H FOOTPRINT_THRESHOLD, --footprint-threshold FOOTPRINT_THRESHOLD
                        When checking test case footprint sizes, warn the user
                        if the new app size is greater then the specified
                        percentage from the last release. Default is 5. 0 to
                        warn on any increase on app size
  -D, --all-deltas      Show all footprint deltas, positive or negative.
                        Implies --footprint-threshold=0
  -O OUTDIR, --outdir OUTDIR
                        Output directory for logs and binaries. Default is
                        'sanity-out' in the current directory. This directory
                        will be deleted unless '--no-clean' is set.
  -n, --no-clean        Do not delete the outdir before building. Will result
                        in faster compilation since builds will be incremental
  -A BOARD_ROOT, --board-root BOARD_ROOT
                        Directory to search for board configuration files. All
                        .yaml files in the directory will be processed.
  -z SIZE, --size SIZE  Don't run sanity checks. Instead, produce a report to
                        stdout detailing RAM/ROM sizes on the specified
                        filenames. All other command line arguments ignored.
  -S, --enable-slow     Execute time-consuming test cases that have been
                        marked as 'slow' in testcase.yaml. Normally these are
                        only built.
  --disable-unrecognized-section-test
                        Skip the 'unrecognized section' test.
  -R, --enable-asserts  deprecated, left for compatibility
  --disable-asserts     deprecated, left for compatibility
  -Q, --error-on-deprecations
                        Error on deprecation warnings.
  --disable-size-report
                        Skip expensive computation of ram/rom segment sizes.
  -x EXTRA_ARGS, --extra-args EXTRA_ARGS
                        Extra CMake cache entries to define when building test
                        cases. May be called multiple times. The key-value
                        entries will be prefixed with -D before being passed
                        to CMake. E.g "sanitycheck -x=USE_CCACHE=0" will
                        translate to "cmake -DUSE_CCACHE=0" which will
                        ultimately disable ccache.
  --west-flash [WEST_FLASH]
                        Uses west instead of ninja or make to flash when
                        running with --device-testing. E.g "sanitycheck
                        --device-testing --device-serial /dev/ttyACM0 --west-
                        flash="--board-id=foobar" will translate to "west
                        flash -- --board-id=foobar" NOTE: device-testing must
                        be enabled to use this option.
  --west-runner WEST_RUNNER
                        Uses the specified west runner instead of default when
                        running with --west-flash. E.g "sanitycheck --device-
                        testing --device-serial /dev/ttyACM0 --west-flash
                        --west-runner=pyocd" will translate to "west flash
                        --runner pyocd" NOTE: west-flash must be enabled to
                        use this option.
  --gcov-tool GCOV_TOOL
                        Path to the gcov tool to use for code coverage reports
  --enable-coverage     Enable code coverage using gcov.
  -C, --coverage        Generate coverage reports. Implies --enable_coverage
                        and --enable-slow
  --coverage-platform COVERAGE_PLATFORM
                        Plarforms to run coverage reports on. This option may
                        be used multiple times. Default to what was selected
                        with --platform.

Test case selection:
  
  Artificially long but functional example:
      $ ./scripts/sanitycheck -v     \
        --testcase-root tests/ztest/base    \
        --testcase-root tests/kernel   \
        --test      tests/ztest/base/testing.ztest.verbose_0  \
        --test      tests/kernel/fifo/fifo_api/kernel.fifo.poll
  
     "kernel.fifo.poll" is one of the test section names in
                                   __/fifo_api/testcase.yaml
      

  -f, --only-failed     Run only those tests that failed the previous sanity
                        check invocation.
  -s TEST, --test TEST  Run only the specified test cases. These are named by 
                        <path/relative/to/Zephyr/base/section.name.in.testcase
                        .yaml>
  --sub-test SUB_TEST   Recursively find sub-test functions and run the entire
                        test section where they were found, including all
                        sibling test functions. Sub-tests are named by: sectio
                        n.name.in.testcase.yaml.function_name_without_test_pre
                        fix Example: kernel.fifo.poll.fifo_loop
  --list-tests          List of all sub-test functions recursively found in
                        all --testcase-root arguments. Note different sub-
                        tests can share the same section name and come from
                        different directories. The output is flattened and
                        reports --sub-test names only, not their directories.
                        For instance net.socket.getaddrinfo_ok and
                        net.socket.fd_set belong to different directories.
  -F FILENAME, --load-tests FILENAME
                        Load list of tests and platforms to be run from file.
  -E FILENAME, --save-tests FILENAME
                        Append list of tests and platforms to be run to file.
  -T TESTCASE_ROOT, --testcase-root TESTCASE_ROOT
                        Base directory to recursively search for test cases.
                        All testcase.yaml files under here will be processed.
                        May be called multiple times. Defaults to the
                        'samples/' and 'tests/' directories at the base of the
                        Zephyr tree.
```

If I add a "West related options" argument group:
```
$ sanitycheck --help
usage: sanitycheck [-h] [--force-toolchain] [-p PLATFORM] [-a ARCH] [-t TAG]
                   [-e EXCLUDE_TAG] [-f] [-c CONFIG]
                   [-s TEST | --sub-test SUB_TEST] [-l] [-o TESTCASE_REPORT]
                   [-d DISCARD_REPORT] [--compare-report COMPARE_REPORT]
                   [-B SUBSET] [-N] [-y] [--list-tags] [--report-excluded]
                   [--list-tests] [--export-tests FILENAME]
                   [--detailed-report FILENAME] [--timestamps] [-r] [-w] [-v]
                   [-i] [--log-file FILENAME] [-m] [-u] [-F FILENAME]
                   [-E FILENAME] [-b | --test-only] [--cmake-only] [-j JOBS]
                   [--device-testing] [-X FIXTURE]
                   [--device-serial DEVICE_SERIAL] [--show-footprint]
                   [-H FOOTPRINT_THRESHOLD] [-D] [-O OUTDIR] [-n]
                   [-T TESTCASE_ROOT] [-A BOARD_ROOT] [-z SIZE] [-S]
                   [--disable-unrecognized-section-test] [-R]
                   [--disable-asserts] [-Q] [--disable-size-report]
                   [-x EXTRA_ARGS] [--west-flash [WEST_FLASH]]
                   [--west-runner WEST_RUNNER] [--gcov-tool GCOV_TOOL]
                   [--enable-coverage] [-C]
                   [--coverage-platform COVERAGE_PLATFORM]

Zephyr Sanity Tests

Also check the "User and Developer Guides" at https://docs.zephyrproject.org/

This script scans for the set of unit test applications in the git
repository and attempts to execute them. By default, it tries to
build each test case on one platform per architecture, using a precedence
list defined in an architecture configuration file, and if possible
run the tests in the QEMU emulator.

Test cases are detected by the presence of a 'testcase.yaml' or a sample.yaml
files in the application's project directory. This file may contain one or more
blocks, each identifying a test scenario. The title of the block is a name for
the test case, which only needs to be unique for the test cases specified in
that testcase meta-data. The full canonical name for each test case is <path to
test case>/<block>.

Each test block in the testcase meta data can define the following key/value
pairs:

  tags: <list of tags> (required)
    A set of string tags for the testcase. Usually pertains to
    functional domains but can be anything. Command line invocations
    of this script can filter the set of tests to run based on tag.

  skip: <True|False> (default False)
    skip testcase unconditionally. This can be used for broken tests.

  slow: <True|False> (default False)
    Don't build or run this test case unless --enable-slow was passed
    in on the command line. Intended for time-consuming test cases
    that are only run under certain circumstances, like daily
    builds.

  extra_args: <list of extra arguments>
    Extra cache entries to pass to CMake when building or running the
    test case.

  extra_configs: <list of extra configurations>
    Extra configuration options to be merged with a master prj.conf
    when building or running the test case.

  build_only: <True|False> (default False)
    If true, don't try to run the test even if the selected platform
    supports it.

  build_on_all: <True|False> (default False)
    If true, attempt to build test on all available platforms.

  depends_on: <list of features>
    A board or platform can announce what features it supports, this option
    will enable the test only those platforms that provide this feature.

  min_ram: <integer>
    minimum amount of RAM needed for this test to build and run. This is
    compared with information provided by the board metadata.

  min_flash: <integer>
    minimum amount of ROM needed for this test to build and run. This is
    compared with information provided by the board metadata.

  timeout: <number of seconds>
    Length of time to run test in QEMU before automatically killing it.
    Default to 60 seconds.

  arch_whitelist: <list of arches, such as x86, arm, arc>
    Set of architectures that this test case should only be run for.

  arch_exclude: <list of arches, such as x86, arm, arc>
    Set of architectures that this test case should not run on.

  platform_whitelist: <list of platforms>
    Set of platforms that this test case should only be run for.

  platform_exclude: <list of platforms>
    Set of platforms that this test case should not run on.

  extra_sections: <list of extra binary sections>
    When computing sizes, sanitycheck will report errors if it finds
    extra, unexpected sections in the Zephyr binary unless they are named
    here. They will not be included in the size calculation.

  filter: <expression>
    Filter whether the testcase should be run by evaluating an expression
    against an environment containing the following values:

    { ARCH : <architecture>,
      PLATFORM : <platform>,
      <all CONFIG_* key/value pairs in the test's generated defconfig>,
      <all DT_* key/value pairs in the test's generated device tree file>,
      <all CMake key/value pairs in the test's generated CMakeCache.txt file>,
      *<env>: any environment variable available
    }

    The grammar for the expression language is as follows:

    expression ::= expression "and" expression
                 | expression "or" expression
                 | "not" expression
                 | "(" expression ")"
                 | symbol "==" constant
                 | symbol "!=" constant
                 | symbol "<" number
                 | symbol ">" number
                 | symbol ">=" number
                 | symbol "<=" number
                 | symbol "in" list
                 | symbol ":" string
                 | symbol

    list ::= "[" list_contents "]"

    list_contents ::= constant
                    | list_contents "," constant

    constant ::= number
               | string

    For the case where expression ::= symbol, it evaluates to true
    if the symbol is defined to a non-empty string.

    Operator precedence, starting from lowest to highest:

        or (left associative)
        and (left associative)
        not (right associative)
        all comparison operators (non-associative)

    arch_whitelist, arch_exclude, platform_whitelist, platform_exclude
    are all syntactic sugar for these expressions. For instance

        arch_exclude = x86 arc

    Is the same as:

        filter = not ARCH in ["x86", "arc"]

    The ':' operator compiles the string argument as a regular expression,
    and then returns a true value only if the symbol's value in the environment
    matches. For example, if CONFIG_SOC="quark_se" then

        filter = CONFIG_SOC : "quark.*"

    Would match it.

The set of test cases that actually run depends on directives in the testcase
filed and options passed in on the command line. If there is any confusion,
running with -v or --discard-report can help show why particular test cases
were skipped.

Metrics (such as pass/fail state and binary size) for the last code
release are stored in scripts/sanity_chk/sanity_last_release.csv.
To update this, pass the --all --release options.

To load arguments from a file, write '+' before the file name, e.g.,
+file_name. File content must be one or more valid arguments separated by
line break instead of white spaces.

Most everyday users will run with no arguments.

optional arguments:
  -h, --help            show this help message and exit
  --force-toolchain     Do not filter based on toolchain, use the set
                        toolchain unconditionally
  -p PLATFORM, --platform PLATFORM
                        Platform filter for testing. This option may be used
                        multiple times. Testcases will only be built/run on
                        the platforms specified. If this option is not used,
                        then platforms marked as default in the platform
                        metadata file will be chosen to build and test.
  -a ARCH, --arch ARCH  Arch filter for testing. Takes precedence over
                        --platform. If unspecified, test all arches. Multiple
                        invocations are treated as a logical 'or' relationship
  -t TAG, --tag TAG     Specify tags to restrict which tests to run by tag
                        value. Default is to not do any tag filtering.
                        Multiple invocations are treated as a logical 'or'
                        relationship
  -e EXCLUDE_TAG, --exclude-tag EXCLUDE_TAG
                        Specify tags of tests that should not run. Default is
                        to run all tests with all tags.
  -c CONFIG, --config CONFIG
                        Specify platform configuration values filtering. This
                        can be specified two ways: <config>=<value> or just
                        <config>. The defconfig for all platforms will be
                        checked. For the <config>=<value> case, only match
                        defconfig that have that value defined. For the
                        <config> case, match defconfig that have that value
                        assigned to any value. Prepend a '!' to invert the
                        match.
  -l, --all             Build/test on all platforms. Any --platform arguments
                        ignored.
  -o TESTCASE_REPORT, --testcase-report TESTCASE_REPORT
                        Output a CSV spreadsheet containing results of the
                        test run. The handler_time column is left blank for
                        tests that were only compiled and not run.
  -d DISCARD_REPORT, --discard-report DISCARD_REPORT
                        Output a CSV spreadsheet showing tests that were
                        skipped and why
  --compare-report COMPARE_REPORT
                        Use this report file for size comparison
  -B SUBSET, --subset SUBSET
                        Only run a subset of the tests, 1/4 for running the
                        first 25%, 3/5 means run the 3rd fifth of the total.
                        This option is useful when running a large number of
                        tests on different hosts to speed up execution time.
  -N, --ninja           Use the Ninja generator with CMake
  -y, --dry-run         Create the filtered list of test cases, but don't
                        actually run them. Useful if you're just interested in
                        --discard-report
  --list-tags           list all tags in selected tests
  --report-excluded     List all tests that are never run based on current
                        scope and coverage. If you are looking for accurate
                        results, run this with --all, but this will take a
                        while...
  --export-tests FILENAME
                        Export tests case meta-data to a file in CSV format.
  --detailed-report FILENAME
                        Generate a junit report with detailed testcase
                        results. Unlike the CSV file produced by --testcase-
                        report, this XML report includes only tests which have
                        run and none which were merely built. If an image with
                        multiple tests crashes early then later tests are not
                        accounted for either.
  --timestamps          Print all messages with time stamps
  -r, --release         Update the benchmark database with the results of this
                        test run. Intended to be run by CI when tagging an
                        official release. This database is used as a basis for
                        comparison when looking for deltas in metrics such as
                        footprint
  -w, --warnings-as-errors
                        Treat warning conditions as errors
  -v, --verbose         Emit debugging information, call multiple times to
                        increase verbosity
  -i, --inline-logs     Upon test failure, print relevant log data to stdout
                        instead of just a path to it
  --log-file FILENAME   log also to file
  -m, --last-metrics    Instead of comparing metrics from the last --release,
                        compare with the results of the previous sanity check
                        invocation
  -u, --no-update       do not update the results of the last run of the
                        sanity checks
  -b, --build-only      Only build the code, do not execute any of it in QEMU
  --test-only           Only run device tests with current artifacts, do not
                        build the code
  --cmake-only          Test on device directly. Specify the serial device to
                        use with the --device-serial option.
  -j JOBS, --jobs JOBS  Number of jobs for building, defaults to number of CPU
                        threads, overcommited by factor 2 when --build-only
  --device-testing      Test on device directly. Specify the serial device to
                        use with the --device-serial option.
  -X FIXTURE, --fixture FIXTURE
                        Specify a fixture that a board might support
  --device-serial DEVICE_SERIAL
                        Serial device for accessing the board (e.g.,
                        /dev/ttyACM0)
  --show-footprint      Show footprint statistics and deltas since last
                        release.
  -H FOOTPRINT_THRESHOLD, --footprint-threshold FOOTPRINT_THRESHOLD
                        When checking test case footprint sizes, warn the user
                        if the new app size is greater then the specified
                        percentage from the last release. Default is 5. 0 to
                        warn on any increase on app size
  -D, --all-deltas      Show all footprint deltas, positive or negative.
                        Implies --footprint-threshold=0
  -O OUTDIR, --outdir OUTDIR
                        Output directory for logs and binaries. Default is
                        'sanity-out' in the current directory. This directory
                        will be deleted unless '--no-clean' is set.
  -n, --no-clean        Do not delete the outdir before building. Will result
                        in faster compilation since builds will be incremental
  -A BOARD_ROOT, --board-root BOARD_ROOT
                        Directory to search for board configuration files. All
                        .yaml files in the directory will be processed.
  -z SIZE, --size SIZE  Don't run sanity checks. Instead, produce a report to
                        stdout detailing RAM/ROM sizes on the specified
                        filenames. All other command line arguments ignored.
  -S, --enable-slow     Execute time-consuming test cases that have been
                        marked as 'slow' in testcase.yaml. Normally these are
                        only built.
  --disable-unrecognized-section-test
                        Skip the 'unrecognized section' test.
  -R, --enable-asserts  deprecated, left for compatibility
  --disable-asserts     deprecated, left for compatibility
  -Q, --error-on-deprecations
                        Error on deprecation warnings.
  --disable-size-report
                        Skip expensive computation of ram/rom segment sizes.
  -x EXTRA_ARGS, --extra-args EXTRA_ARGS
                        Extra CMake cache entries to define when building test
                        cases. May be called multiple times. The key-value
                        entries will be prefixed with -D before being passed
                        to CMake. E.g "sanitycheck -x=USE_CCACHE=0" will
                        translate to "cmake -DUSE_CCACHE=0" which will
                        ultimately disable ccache.
  --gcov-tool GCOV_TOOL
                        Path to the gcov tool to use for code coverage reports
  --enable-coverage     Enable code coverage using gcov.
  -C, --coverage        Generate coverage reports. Implies --enable_coverage
                        and --enable-slow
  --coverage-platform COVERAGE_PLATFORM
                        Plarforms to run coverage reports on. This option may
                        be used multiple times. Default to what was selected
                        with --platform.

Test case selection:
  
  Artificially long but functional example:
      $ ./scripts/sanitycheck -v     \
        --testcase-root tests/ztest/base    \
        --testcase-root tests/kernel   \
        --test      tests/ztest/base/testing.ztest.verbose_0  \
        --test      tests/kernel/fifo/fifo_api/kernel.fifo.poll
  
     "kernel.fifo.poll" is one of the test section names in
                                   __/fifo_api/testcase.yaml
      

  -f, --only-failed     Run only those tests that failed the previous sanity
                        check invocation.
  -s TEST, --test TEST  Run only the specified test cases. These are named by 
                        <path/relative/to/Zephyr/base/section.name.in.testcase
                        .yaml>
  --sub-test SUB_TEST   Recursively find sub-test functions and run the entire
                        test section where they were found, including all
                        sibling test functions. Sub-tests are named by: sectio
                        n.name.in.testcase.yaml.function_name_without_test_pre
                        fix Example: kernel.fifo.poll.fifo_loop
  --list-tests          List of all sub-test functions recursively found in
                        all --testcase-root arguments. Note different sub-
                        tests can share the same section name and come from
                        different directories. The output is flattened and
                        reports --sub-test names only, not their directories.
                        For instance net.socket.getaddrinfo_ok and
                        net.socket.fd_set belong to different directories.
  -F FILENAME, --load-tests FILENAME
                        Load list of tests and platforms to be run from file.
  -E FILENAME, --save-tests FILENAME
                        Append list of tests and platforms to be run to file.
  -T TESTCASE_ROOT, --testcase-root TESTCASE_ROOT
                        Base directory to recursively search for test cases.
                        All testcase.yaml files under here will be processed.
                        May be called multiple times. Defaults to the
                        'samples/' and 'tests/' directories at the base of the
                        Zephyr tree.

West related options:
  --west-flash [WEST_FLASH]
                        Uses west instead of ninja or make to flash when
                        running with --device-testing. E.g "sanitycheck
                        --device-testing --device-serial /dev/ttyACM0 --west-
                        flash="--board-id=foobar" will translate to "west
                        flash -- --board-id=foobar" NOTE: device-testing must
                        be enabled to use this option.
  --west-runner WEST_RUNNER
                        Uses the specified west runner instead of default when
                        running with --west-flash. E.g "sanitycheck --device-
                        testing --device-serial /dev/ttyACM0 --west-flash
                        --west-runner=pyocd" will translate to "west flash
                        --runner pyocd" NOTE: west-flash must be enabled to
                        use this option.
```

I'm sorry, but the west options now look like orphans who have lost their way.  I understand that I could come up with "creative naming" to make them sort a bit better -- and this is the naming discussion that I felt might be needed to do this correctly.

IMHO, adding groups only works if we do it for all of the options.

mike-scott commented:
@marc-hb ^^

marc-hb commented:
The important thing IMHO is that the two options closely related to each other appear next to each other. Thanks to alphabetical order this happens in either case so even if using a group looks marginally better either is fine by me.
